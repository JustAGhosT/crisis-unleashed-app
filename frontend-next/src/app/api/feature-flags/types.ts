// Strongly-typed feature flags
export const FLAG_KEYS = [
  "useNewFactionUI",
  "useNewDeckBuilder",
  "useNewCardDisplay",
  "useNewNavigation",
  "useNewTheme",
] as const;

export type FlagKey = (typeof FLAG_KEYS)[number];

export interface FeatureFlags {
  useNewFactionUI: boolean;
  useNewDeckBuilder: boolean;
  useNewCardDisplay: boolean;
  useNewNavigation: boolean;
  useNewTheme: boolean;
}
