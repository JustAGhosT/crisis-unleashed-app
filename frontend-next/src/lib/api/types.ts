/**
 * Comprehensive TypeScript types for API responses with runtime validation
 */

import { z } from "zod";

// Base response schema
export const BaseResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

// User schema
export const UserSchema = z.object({
  id: z.string(),
  username: z.string(),
  email: z.string().email().optional(),
  name: z.string().optional(),
  avatar: z.string().url().optional(),
  role: z.enum(["user", "admin", "moderator"]).default("user"),
});

// Authentication response schema
export const AuthResponseSchema = BaseResponseSchema.extend({
  user: UserSchema.optional(),
  token: z.string().optional(),
});

// Card schema
export const CardSchema = z.object({
  id: z.string(),
  name: z.string(),
  faction: z.enum([
    "solaris_nexus",
    "umbral_eclipse",
    "aeonic_dominion",
    "primordial_genesis",
    "infernal_core",
    "neuralis_conclave",
    "synthetic_directive"
  ]),
  type: z.enum(["hero", "unit", "action", "structure"]),
  cost: z.number().min(0),
  rarity: z.enum(["common", "uncommon", "rare", "epic", "legendary"]),
  description: z.string(),
  abilities: z.array(z.string()).default([]),
  stats: z.object({
    attack: z.number().min(0).optional(),
    health: z.number().min(1).optional(),
    defense: z.number().min(0).optional(),
  }).optional(),
});

// Deck schema
export const DeckSchema = z.object({
  id: z.string(),
  name: z.string(),
  faction_primary: CardSchema.shape.faction,
  faction_secondary: CardSchema.shape.faction.optional(),
  cards: z.array(z.object({
    card_id: z.string(),
    quantity: z.number().min(1).max(3),
  })),
  hero_id: z.string(),
  is_public: z.boolean().default(false),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

// API response schemas
export const CardListResponseSchema = BaseResponseSchema.extend({
  cards: z.array(CardSchema),
  total: z.number().optional(),
  page: z.number().optional(),
  limit: z.number().optional(),
});

export const DeckListResponseSchema = BaseResponseSchema.extend({
  decks: z.array(DeckSchema),
  total: z.number().optional(),
});

export const HealthCheckResponseSchema = z.object({
  status: z.enum(["healthy", "degraded", "unhealthy"]),
  version: z.string(),
  timestamp: z.string().datetime(),
  services: z.record(z.object({
    status: z.enum(["up", "down", "degraded"]),
    response_time: z.number().optional(),
    last_check: z.string().datetime().optional(),
  })),
});

// Type exports
export type BaseResponse = z.infer<typeof BaseResponseSchema>;
export type User = z.infer<typeof UserSchema>;
export type AuthResponse = z.infer<typeof AuthResponseSchema>;
export type Card = z.infer<typeof CardSchema>;
export type Deck = z.infer<typeof DeckSchema>;
export type CardListResponse = z.infer<typeof CardListResponseSchema>;
export type DeckListResponse = z.infer<typeof DeckListResponseSchema>;
export type HealthCheckResponse = z.infer<typeof HealthCheckResponseSchema>;

// Validation helper functions
export function validateApiResponse<T>(schema: z.ZodSchema<T>, data: unknown): T {
  try {
    return schema.parse(data);
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error("API Response validation failed:", error.errors);
      throw new Error(`Invalid API response format: ${error.errors.map(e => e.message).join(', ')}`);
    }
    throw error;
  }
}

// Safe parsing with defaults
export function safeParseApiResponse<T>(
  schema: z.ZodSchema<T>,
  data: unknown,
  defaultValue: T
): T {
  const result = schema.safeParse(data);
  if (result.success) {
    return result.data;
  }

  console.warn("API Response validation failed, using default:", result.error.errors);
  return defaultValue;
}

// Request validation schemas
export const LoginRequestSchema = z.object({
  username: z.string().optional(),
  email: z.string().email().optional(),
  password: z.string().min(1),
}).refine(data => data.username || data.email, {
  message: "Either username or email must be provided",
});

export const CreateDeckRequestSchema = z.object({
  name: z.string().min(1).max(50),
  faction_primary: CardSchema.shape.faction,
  faction_secondary: CardSchema.shape.faction.optional(),
  hero_id: z.string(),
  cards: z.array(z.object({
    card_id: z.string(),
    quantity: z.number().min(1).max(3),
  })).min(30).max(50),
  is_public: z.boolean().default(false),
});

export type LoginRequest = z.infer<typeof LoginRequestSchema>;
export type CreateDeckRequest = z.infer<typeof CreateDeckRequestSchema>;