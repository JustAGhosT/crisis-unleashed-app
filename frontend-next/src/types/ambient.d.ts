// Ambient declarations placeholder. Intentionally left minimal.
// Note: jsonwebtoken and web-vitals are now typed via installed packages.
