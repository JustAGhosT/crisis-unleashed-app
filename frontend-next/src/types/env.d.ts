// Ambient type declarations for environment variables used in the app
// Ensures process.env access is type-safe and unified across the codebase

export {};

// eslint-disable-next-line @typescript-eslint/no-namespace
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace NodeJS {
    interface ProcessEnv {
      // Allow arbitrary environment variables while keeping explicit ones typed
      readonly [key: string]: string | undefined;
      readonly ADMIN_TOKEN?: string;
      readonly NEXT_PUBLIC_API_URL?: string;
      // Feature flag envs (server-evaluated booleans as strings)
      readonly ENABLE_NEW_FACTION_UI?: string;
      readonly ENABLE_NEW_DECK_BUILDER?: string;
      readonly ENABLE_NEW_CARD_DISPLAY?: string;
      readonly ENABLE_NEW_NAVIGATION?: string;
      readonly ENABLE_NEW_THEME?: string;
    }
  }
}
