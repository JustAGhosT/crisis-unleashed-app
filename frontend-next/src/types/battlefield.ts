import { Axial } from '@/lib/hex';
import { BattlefieldUnit, BattlefieldZone } from '@/types/game';

export type MovementCostFn = (
  unit: BattlefieldUnit, 
  src: BattlefieldZone, 
  dst: BattlefieldZone, 
  dist: number
) => number;

// Extend BattlefieldUnit with required properties from game-algorithms.md
export interface EnhancedBattlefieldUnit extends BattlefieldUnit {
  initiative?: number;
  ethereal?: boolean;
  zocCostModifier?: number;
  abilities?: string[];
  shields?: number;
  guards?: number;
  statusEffects?: StatusEffect[];
}

export interface StatusEffect {
  type: string;
  duration: number;
  stacks?: number;
  exclusive?: boolean;
}

export enum CombatTiming {
  PRE_ATTACK = "pre-attack",
  ON_ATTACK = "on-attack", 
  ON_HIT = "on-hit",
  POST_HIT = "post-hit",
  ON_KILL = "on-kill",
  END_OF_COMBAT = "end-of-combat"
}

export interface CombatState {
  attacker: BattlefieldUnit;
  defender: BattlefieldUnit;
  damage: number;
  canceled: boolean;
  lethal: boolean;
}

export enum Phase {
  START_OF_TURN = "start-of-turn",
  RESOURCE_GAIN = "resource-gain",
  DEPLOY = "deploy",
  ACTION = "action",
  END = "end",
  CLEANUP = "cleanup"
}