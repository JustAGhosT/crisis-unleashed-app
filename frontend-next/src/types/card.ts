import { FactionId } from "./faction";

export type CardType = "hero" | "unit" | "action" | "structure";
export type UnitType = "melee" | "ranged" | "siege" | "flying";
export type ActionType = "instant" | "ongoing" | "equipment";
export type StructureType = "building" | "trap" | "aura";
export type CardRarity = "common" | "uncommon" | "rare" | "epic" | "legendary";

export interface Card {
  id: string;
  name: string;
  description: string;
  type: CardType;
  unitType?: UnitType; // Required for unit cards
  actionType?: ActionType; // Required for action cards
  structureType?: StructureType; // Required for structure cards
  faction: FactionId;
  rarity: CardRarity;
  cost: number;
  attack?: number; // Required for unit/hero cards
  health?: number; // Required for unit/hero cards
  initiative?: number; // For combat sequencing
  range?: number; // For ranged units and line of sight
  movementSpeed?: number; // For battlefield positioning
  keywords?: string[]; // Combat keywords: Quick Attack, Overwhelm, etc.
  abilities: string[];
  playConditions?: string[]; // Play condition verification
  energyCost: number; // Energy system integration
  momentumCost?: number; // Momentum system integration
  zoneRestrictions?: string[]; // Zone-based battlefield rules
  persistsOnBattlefield?: boolean; // For structure cards
  imageUrl?: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface DeckCard {
  cardId: string;
  quantity: number;
}

export interface Deck {
  id: string;
  userId: string;
  name: string;
  faction: FactionId;
  isActive: boolean;
  cards: DeckCard[];
  createdAt: string;
  updatedAt: string;
}

export interface UserCard {
  id: string;
  userId: string;
  cardId: string;
  quantity: number;
  isFavorite: boolean;
  acquiredAt: string;
  nftId?: string;
}

// Card filtering and search
export interface CardFilters {
  faction?: FactionId;
  type?: CardType;
  rarity?: CardRarity;
  costMin?: number;
  costMax?: number;
  search?: string;
  abilities?: string[];
  owned?: boolean;
}

export interface CardSearchResult {
  cards: Card[];
  total: number;
  page: number;
  pageSize: number;
}

// Deck validation
export interface DeckValidationResult {
  isValid: boolean;
  errors: string[];
  warnings: string[];
  cardCount: number;
  heroCardCount: number;
  unitCardCount: number;
  actionCardCount: number;
  structureCardCount: number;
  factionConsistency: boolean;
  energyCurveBalance: boolean;
  costCurve: { [cost: number]: number };
  factionSpecificRules: { [rule: string]: boolean };
}

export interface DeckStats {
  totalCards: number;
  averageCost: number;
  averageInitiative: number;
  frontlineUnitCount: number;
  backlineUnitCount: number;
  rangedUnitCount: number;
  flyingUnitCount: number;
  typeDistribution: Record<CardType, number>;
  unitTypeDistribution: Record<UnitType, number>;
  actionTypeDistribution: Record<ActionType, number>;
  structureTypeDistribution: Record<StructureType, number>;
  rarityDistribution: Record<CardRarity, number>;
  costCurve: { [cost: number]: number };
  energyCurve: { [cost: number]: number };
  momentumRequirements: { [cost: number]: number };
}
