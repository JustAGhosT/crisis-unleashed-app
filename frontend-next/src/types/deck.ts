/**
 * Types for the deck builder system
 */

export interface Card {
  id: string;
  instanceId?: string; // Unique ID for each instance of a card in a deck
  name: string;
  cost?: number;
  power?: number;
  type?: string;
  subtype?: string;
  faction?: string;
  rarity?: string;
  text?: string;
  imageUrl?: string;
  isUnique?: boolean;
  keywords?: string[];
}

export interface Deck {
  id?: string;
  name: string;
  description?: string;
  faction?: string;
  cards: Card[];
  maxCards?: number;
  minCards?: number;
  createdAt?: string;
  updatedAt?: string;
  author?: string;
}

export interface DeckValidationError {
  code: string;
  message: string;
}

export interface DeckStats {
  cardCount: number;
  uniqueCardCount: number;
  averageCost: number;
  averagePower: number;
  typeDistribution: Record<string, number>;
  costDistribution: Record<number, number>;
  rarityDistribution: Record<string, number>;
}

export interface DeckFilters {
  name?: string;
  faction?: string;
  author?: string;
  sortBy?: "name" | "created" | "updated" | "popularity";
  sortOrder?: "asc" | "desc";
}

export interface CardFilters {
  name?: string;
  type?: string;
  subtype?: string;
  faction?: string;
  cost?: number | [number, number]; // Either exact cost or range
  rarity?: string;
  keyword?: string;
}

// --- Deck Builder UI additions ---
export interface DeckRowVM {
  cardId: string;
  quantity: number;
  name: string;
  cost?: number;
  type?: string;
  faction?: string;
  rarity?: string;
  attack?: number;
  health?: number;
}

export const CARD_ID_MIME = "text/card-id";
export const DECK_CARD_ID_MIME = "text/deck-card-id";
export const CARD_JSON_MIME = "application/json";
