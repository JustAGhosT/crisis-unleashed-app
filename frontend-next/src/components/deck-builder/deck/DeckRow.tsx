import React from "react";
import { Button } from "@/components/ui/button";
import { Badge } from "@/components/ui/badge";
import { Trash2 } from "lucide-react";
import { cn } from "@/lib/utils";
import { rarityBadgeClass } from "@/lib/ui-maps";
import { Card as GameCardData } from "@/types/card";

interface DeckRowProps {
  card: GameCardData;
  quantity: number;
  selected?: boolean;
  disabled?: boolean;
  draggable?: boolean;
  onAdd: (card: GameCardData) => void;
  onRemove: (card: GameCardData) => void;
  onClick?: (card: GameCardData) => void;
  onDragStart?: (e: React.DragEvent) => void;
  onDragOver?: (e: React.DragEvent) => void;
  onDrop?: (e: React.DragEvent) => void;
  onKeyDown?: (e: React.KeyboardEvent) => void;
}

const DeckRowInner: React.FC<DeckRowProps> = ({
  card,
  quantity,
  selected = false,
  disabled = false,
  onAdd,
  onRemove,
  onClick,
  draggable = false,
  onDragStart,
  onDragOver,
  onDrop,
  onKeyDown,
}) => {
  // Enable the native draggable attribute only during pointer interaction
  const [dragAttrEnabled, setDragAttrEnabled] = React.useState(false);
  const enableDragAttr = React.useCallback(() => {
    if (draggable) setDragAttrEnabled(true);
  }, [draggable]);
  const disableDragAttr = React.useCallback(() => {
    if (dragAttrEnabled) setDragAttrEnabled(false);
  }, [dragAttrEnabled]);

  const rootClasses = cn(
    "flex items-center justify-between bg-slate-800/50 border border-slate-600 rounded-lg p-3 hover:bg-slate-800",
    selected && "ring-2 ring-amber-400 ring-offset-2 ring-offset-slate-900",
  );

  return (
    <div
      data-testid="deck-row"
      className={rootClasses}
      draggable={draggable ? true : false}
      onClick={() => onClick?.(card)}
      onDragStart={onDragStart}
      onDragOver={onDragOver}
      onDrop={onDrop}
      onKeyDownCapture={onKeyDown}
      onPointerDown={enableDragAttr}
      onPointerUp={disableDragAttr}
      onPointerCancel={disableDragAttr}
      onMouseDown={enableDragAttr}
      onMouseUp={disableDragAttr}
      onBlur={disableDragAttr}
      tabIndex={0}
      role="listitem"
    >
      {/* Drag handle slot via unicode when draggable */}
      {draggable && (
        <div
          className="cursor-grab mr-2 select-none"
          title="Drag to reorder"
          aria-hidden
        >
          ≡
        </div>
      )}
      {/* Quantity */}
      <div className="flex items-center gap-1">
        <Button
          size="sm"
          variant="outline"
          onClick={(e) => {
            e.stopPropagation();
            onRemove(card);
          }}
          disabled={disabled}
          className="h-6 w-6 p-0"
          aria-label={`Remove one ${card.name}`}
        >
          -
        </Button>
        <span className="w-8 text-center text-sm font-medium">{quantity}</span>
        <Button
          size="sm"
          variant="outline"
          onClick={(e) => {
            e.stopPropagation();
            onAdd(card);
          }}
          disabled={disabled || quantity >= 3}
          className="h-6 w-6 p-0"
          aria-label={`Add one ${card.name}`}
        >
          +
        </Button>
      </div>

      {/* Card details */}
      <div className="flex-1 px-3">
        <div className="font-semibold text-white">{card.name}</div>
        <div className="text-xs text-gray-400">
          Cost: {card.cost ?? 0}
          {card.rarity && (
            <Badge className={cn("ml-2", rarityBadgeClass(card.rarity))}>
              {card.rarity}
            </Badge>
          )}
        </div>
        <div className="text-xs text-gray-400">
          {card.type} • {card.faction}
          {card.attack !== undefined && card.health !== undefined && (
            <span>
              {" "}
              • {card.attack}/{card.health}
            </span>
          )}
        </div>
      </div>

      {/* Remove Button */}
      <Button
        size="sm"
        variant="ghost"
        onClick={(e) => {
          e.stopPropagation();
          onRemove(card);
        }}
        disabled={disabled}
        className="text-red-400 hover:text-red-300 hover:bg-red-400/10"
        aria-label={`Remove ${card.name} from deck`}
      >
        <Trash2 className="w-4 h-4" />
      </Button>
    </div>
  );
};

export const DeckRow = React.memo(DeckRowInner);
DeckRow.displayName = "DeckRow";
