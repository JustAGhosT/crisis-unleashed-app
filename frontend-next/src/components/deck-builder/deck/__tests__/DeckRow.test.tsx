import React from "react";
import { render } from "@testing-library/react";
import { DeckRow } from "../DeckRow";
import { Card as GameCardData } from "@/types/card";

const card: GameCardData = {
  id: "c1",
  name: "Alpha",
  description: "Test unit",
  type: "unit",
  faction: "solaris",
  rarity: "common",
  cost: 2,
  abilities: [],
  energyCost: 2,
  isActive: true,
  createdAt: new Date(0).toISOString(),
  updatedAt: new Date(0).toISOString(),
};

describe("DeckRow", () => {
  it("renders and matches snapshot", () => {
    const { container } = render(
      <DeckRow
        card={card}
        quantity={2}
        onAdd={jest.fn()}
        onRemove={jest.fn()}
        draggable
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
