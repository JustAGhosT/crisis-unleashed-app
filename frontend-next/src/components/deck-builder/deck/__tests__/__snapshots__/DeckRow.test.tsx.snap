// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckRow renders and matches snapshot 1`] = `
<div
  class="flex items-center justify-between bg-slate-800/50 border border-slate-600 rounded-lg p-3 hover:bg-slate-800"
  data-testid="deck-row"
  draggable="true"
  tabindex="0"
>
  <div
    aria-hidden="true"
    class="cursor-grab mr-2 select-none"
    title="Drag to reorder"
  >
    ≡
  </div>
  <div
    class="flex items-center gap-1"
  >
    <button
      aria-label="Remove one Alpha"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3 h-6 w-6 p-0"
    >
      -
    </button>
    <span
      class="w-8 text-center text-sm font-medium"
    >
      2
    </span>
    <button
      aria-label="Add one Alpha"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3 h-6 w-6 p-0"
    >
      +
    </button>
  </div>
  <div
    class="flex-1 px-3"
  >
    <div
      class="font-semibold text-white"
    >
      Alpha
    </div>
    <div
      class="text-xs text-gray-400"
    >
      Cost: 
      2
      <div
        class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent hover:bg-primary/80 ml-2 bg-gray-100 text-gray-800 dark:bg-gray-800/40 dark:text-gray-300"
      >
        common
      </div>
    </div>
    <div
      class="text-xs text-gray-400"
    >
      unit
       • 
      solaris
    </div>
  </div>
  <button
    aria-label="Remove Alpha from deck"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3 text-red-400 hover:text-red-300 hover:bg-red-400/10"
  >
    <svg
      class="lucide lucide-trash2 w-4 h-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 6h18"
      />
      <path
        d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
      />
      <path
        d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
      />
      <line
        x1="10"
        x2="10"
        y1="11"
        y2="17"
      />
      <line
        x1="14"
        x2="14"
        y1="11"
        y2="17"
      />
    </svg>
  </button>
</div>
`;
